dipy/reconst/tests/test_ivim.py

```diff
@@ -0,0 +1,175 @@
+    assert_array_almost_equal(est_signal, data_single)
+    assert_array_almost_equal(ivim_fit.model_params, params)
+
+    # Test predict function for single voxel
+    p = ivim_fit.predict(gtab)
+    assert_array_equal(p.shape, data_single.shape)
+    assert_array_almost_equal(p, data_single)
+
+
+def test_multivoxel():
+    """Test fitting with multivoxel data.
+
+    We generate a multivoxel signal to test the fitting for multivoxel data.
+    This is to ensure that the fitting routine takes care of signals packed as
+    1D, 2D or 3D arrays.
+    """
+    ivim_model = IvimModel(gtab)
+    ivim_fit = ivim_model.fit(data_multi)
+
+    est_signal = ivim_fit.predict(gtab, S0=1.)
+    assert_array_equal(est_signal.shape, data_multi.shape)
+    assert_array_almost_equal(est_signal, data_multi)
+    assert_array_almost_equal(ivim_fit.model_params, ivim_params)
+
+
+def test_ivim_errors():
+    """
+    Test if errors raised in the module are working correctly.
+
+    Scipy introduced bounded least squares fitting in the version 0.17
+    and is not supported by the older versions. Initializing an IvimModel
+    with bounds for older Scipy versions should raise an error.
+    """
+    # Run the test for Scipy versions less than 0.17
+    if SCIPY_VERSION < '0.17':
+        assert_raises(ValueError, IvimModel, gtab,
+                      bounds=([0., 0., 0., 0.], [np.inf, 1., 1., 1.]))
+    else:
+        ivim_model = IvimModel(gtab,
+                               bounds=([0., 0., 0., 0.], [np.inf, 1., 1., 1.]))
+
+
+def test_mask():
+    """
+    Test whether setting incorrect mask raises and error
+    """
+    ivim_model = IvimModel(gtab)
+    mask_correct = data_multi[..., 0] > 0.2
+    mask_not_correct = np.array([[False, True, False], [True, False]])
+
+    fit = ivim_model.fit(data_multi, mask_correct)
+    assert_raises(ValueError, ivim_model.fit, data_multi,
+                  mask=mask_not_correct)
+
+
+def test_with_higher_S0():
+    """
+    Test whether fitting works for S0 > 1.
+    """
+    # params for a single voxel
+    S0_2 = 1000.
+    params2 = np.array([S0_2, f, D_star, D])
+    mevals2 = np.array(([D_star, D_star, D_star], [D, D, D]))
+    # This gives an isotropic signal.
+    signal2 = multi_tensor(gtab, mevals2, snr=None, S0=S0_2,
+                           fractions=[f * 100, 100 * (1 - f)])
+    # Single voxel data
+    data_single2 = signal2[0]
+
+    ivim_model = IvimModel(gtab)
+    ivim_fit = ivim_model.fit(data_single2)
+
+    est_signal = ivim_fit.predict(gtab)
+    assert_array_equal(est_signal.shape, data_single2.shape)
+    assert_array_almost_equal(est_signal, data_single2)
+    assert_array_almost_equal(ivim_fit.model_params, params2)
+
+
+def test_bounds_x0():
+    """
+    Test to check if setting bounds for signal where initial value is
+    higer than subsequent values works.
+    """
+    test_signal = np.array([4574.34814453, 4745.18164062,  4759.51806641, 4618.24951172, 4665.63623047,
```

Comment:
Great. Might be worth adding a comment explaining where these values are taken from (from the IVIM data set, voxel coordinates...).